ansible_base/lib/routers/association_resource_router.py

```diff
@@ -159,6 +162,11 @@ def get_serializer_class(self):
         return super().get_serializer_class()
 
 
+@property
+def attribute_raiser(cls):
+    raise AttributeError
```

Comment:
Is an Attribute Error correct for these? Shouldn't this return a MethodNotAllowed response?

Reply:
Not a 405, a 404. See title, unwanted _URL registrations_, meaning endpoints that shouldn't exist were giving 500 errors. These are endpoints that _no_ method is valid for. If you have the wrong method (GET/POST) but another method would have worked, that's a 405, but that's not what this is fixing. This is _how_ we get that behavior. This is guided missile to programmed to a specific `hasattr` in `SimpleRouter.get_method_map`:

```python
        bound_methods = {}
        for method, action in method_map.items():
            if hasattr(viewset, action):
                bound_methods[method] = action
        return bound_methods
```

I skipped ahead without explaining this in great detail. But this is the way viewsets work, if a method is defined on a viewset, then that's a URL. That method can have multiple valid methods. I need to remove the `retrieve` method and others. If left in, these would create "phantom" endpoints, as in, a detail endpoint nested in the association endpoint, because it got left in the association viewset. So we need to leave it in the original (parent) viewset (because it is used for other things) and remove it as we subclass it.

The unorthodox thing is that `viewset` is a _class_, and we need a _subclass_ that gives `False` for `hasattr(viewset, 'retrieve')`. That is what `AttributeError` does. It's a method-scrubber, isolated to the new subclass.